packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+  typeof controllerName,
+  AllowedActions,
+  any,
+  AllowedActions['type'],
+  any
+>;
+
+type SessionData = {
+  origin: string;
+  requestedNamespaces: Record<NamespaceId, RequestNamespace>;
+  providedNamespaces: Record<NamespaceId, RequestNamespace>;
+  handlingSnaps: Record<NamespaceId, SnapId>;
+};
+
+type MultiChainControllerState = {
+  sessions: { [origin: string]: SessionData };
+};
+
+type Notify = (
+  origin: string,
+  data: { method: string; params?: Record<string, unknown> },
+) => Promise<void>;
+
+type MultiChainControllerArgs = {
+  notify: Notify;
+  messenger: MultiChainControllerMessenger;
+};
+
+/**
+ * Parse a chain ID string to an object containing the namespace and reference.
+ * This validates the chain ID before parsing it.
+ *
+ * @param chainId - The chain ID to validate and parse.
+ * @returns The parsed chain ID.
+ */
+export function parseChainId(chainId: ChainId): {
+  namespace: NamespaceId;
+  reference: string;
+} {
+  if (!isChainId(chainId)) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  const match = CHAIN_ID_REGEX.exec(chainId);
+  if (!match?.groups) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  return {
+    namespace: match.groups.namespace,
+    reference: match.groups.reference,
+  };
+}
+
+/**
+ * Parse an account ID to an object containing the chain, chain ID and address.
+ * This validates the account ID before parsing it.
+ *
+ * @param accountId - The account ID to validate and parse.
+ * @returns The parsed account ID.
+ */
+export function parseAccountId(accountId: AccountId): {
+  chain: { namespace: NamespaceId; reference: string };
+  chainId: ChainId;
+  address: string;
+} {
+  if (!isAccountId(accountId)) {
+    throw new Error('Invalid account ID.');
+  }
+
+  const match = ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress,
+    chainId: match.groups.chainId as ChainId,
+    chain: {
+      namespace: match.groups.namespace,
+      reference: match.groups.reference,
+    },
+  };
+}
+
+// TODO(ritave): Support for legacy ethereum operations, not just snaps
+export class MultiChainController extends BaseController<
+  typeof controllerName,
+  MultiChainControllerState,
+  MultiChainControllerMessenger
+> {
+  private notify: Notify;
+
+  constructor({ messenger, notify }: MultiChainControllerArgs) {
+    super({
+      messenger,
+      metadata: {
+        sessions: { persist: false, anonymous: false },
+      },
+      name: controllerName,
+      state: defaultState,
+    });
+
+    this.notify = notify;
+  }
+
+  getSession(origin: string): SessionData | undefined {
+    return this.state.sessions[origin];
+  }
+
+  async closeSession(origin: string): Promise<void> {
+    const session = this.getSession(origin);
+    assert(session, new Error('No session to close'));
+
+    await this.notify(origin, { method: 'multichainHack_metamask_disconnect' });
```

Comment:
Oh, I was thinking more like `this.notify(origin, { method: 'multichainHack_event', params: { method: 'metamask_disconnect' } });`

But this works as well

Reply:
No strong opinion on this, we can do either.

Reply:
Let's nest it so it's easier to remove using find-all IDE tool

Reply:
Actually this will complicate the provider. I vote we keep as-is. 
https://github.com/MetaMask/snaps-skunkworks/blob/ritave/provider/packages/provider/src/MultiChainProvider.ts#L26-L42

Reply:
That url is outdated, can't see the problem :/

Reply:
Updated URL: https://github.com/MetaMask/snaps-skunkworks/blob/ritave/provider/packages/provider/src/MultiChainProvider.ts#L35-L50

I just meant it complicates the listening of events if we nest the events. We would need 1 listener that switches based on the params. It adds more complexity than its worth imo.